### PR TITLE
Add split kernel

### PR DESCRIPTION
**Pull request by haozech**
This PR mainly did 1 things:
For cuda backends, when input stages contains more than 1 group, split the original kernel into multi kernels.
For example:
```
  Placeholder<float> A("X", {M, N});
  Placeholder<float> B("Y", {M, N});

  auto C = Compute(
      {M, N}, [&](Var i, Var j) { return A(i, j) * B(i, j); }, "C");

  auto D = Compute(
      {M, N}, [&](Var i, Var j) { return C(i, j) + B(i, j); }, "D");

  auto stages = CreateStages({C, D});

  stages[C]->Bind(0, "blockIdx.x");
  stages[C]->Bind(1, "threadIdx.x");
  stages[D]->Bind(0, "blockIdx.x");
  stages[D]->Bind(1, "threadIdx.x");

  CodeGenCUDA_Dev codegen(target);

  auto func = lang::LowerVec("elementwise_add", stages, {A, B, C, D}, {}, {}, nullptr, target);
```
In the past, the func will be 
```
__global__
void elementwise_add(const float* __restrict__ X, const float* __restrict__ Y, float* __restrict__ C)
{
  if ((blockIdx.x < 100)) {
    if ((threadIdx.x < 200)) {
      C[((200 * blockIdx.x) + threadIdx.x)] = (X[((200 * blockIdx.x) + threadIdx.x)] * Y[((200 * blockIdx.x) + threadIdx.x)]);
    };
  };
  if ((blockIdx.x < 100)) {
    if ((threadIdx.x < 200)) {
      D[((200 * blockIdx.x) + threadIdx.x)] = (C[((200 * blockIdx.x) + threadIdx.x)] + Y[((200 * blockIdx.x) + threadIdx.x)]);
    };
  };
}
```

Now it will be split into 2 kernels:
```
__global__
void elementwise_add(const float* __restrict__ X, const float* __restrict__ Y, float* __restrict__ C)
{
  if ((blockIdx.x < 100)) {
    if ((threadIdx.x < 200)) {
      C[((200 * blockIdx.x) + threadIdx.x)] = (X[((200 * blockIdx.x) + threadIdx.x)] * Y[((200 * blockIdx.x) + threadIdx.x)]);
    };
  };
}__global__
void elementwise_add_1(const float* __restrict__ X, const float* __restrict__ Y, const float* __restrict__ C, float* __restrict__ D)
{
  if ((blockIdx.x < 100)) {
    if ((threadIdx.x < 200)) {
      D[((200 * blockIdx.x) + threadIdx.x)] = (C[((200 * blockIdx.x) + threadIdx.x)] + Y[((200 * blockIdx.x) + threadIdx.x)]);
    };
  };
}
```

Note that for prudential reasons, this change doesn't effect X86 backends.

### CODE REVIEW

File: cinn/hlir/framework/instruction.h [hunk 1]
```diff
@@ -37,21 +37,25 @@ class Instruction {
...
   /**
    * Run the Instruction.
    */
   void Run() {
-    CHECK(fn_) << "The LoweredFunc address should be set first by calling SetLoweredFunc method";
-    auto& pod_args = PreparePodArgs();
+    if (fn_.size() > 1 && fn_.size() != in_args_.size()) {
+      out_args_.back()[0] = out_args_.front()[0];
+      out_args_.erase(out_args_.begin());
+      in_args_.erase(in_args_.begin());
```
Comment (wenming2014) [inline cinn/hlir/framework/instruction.h:55]: 这些操作作用是？

File: cinn/hlir/framework/graph_compiler.cc [hunk 2]
```diff
@@ -266,21 +276,92 @@ std::unique_ptr<Program> GraphCompiler::Build(const std::string& code) {
...
   } else {
     VLOG(3) << "not run opfusion pass";
     for (auto& node : nodes) {
       auto op_node = node->safe_as<Node>();
       if (op_node) {
         auto lowered_func = GetOpFunc(op_node);
-        m_builder_.AddFunction(lowered_func);
+
+        if (lowered_func.size() > 1) {
+          for (auto& i : lowered_func) {
```
Comment (wenming2014) [inline cinn/hlir/framework/graph_compiler.cc]: 这块代码重复，可以整合一下吗？

File: cinn/lang/lower_impl.cc [hunk 3]
```diff
@@ -399,6 +402,95 @@ std::vector<ir::Argument> LowerImpl::GenerateFunctionArgumentList(Expr fn_body)
 
   return args;
 }
+// Generate Function Arguments for splitted kernel.
+std::vector<ir::Argument> LowerImpl::GenerateFunctionArgument2(Expr func_iterator,
```
Comment (wenming2014) [inline cinn/lang/lower_impl.cc]: 函数名可以考虑改一下？

File: cinn/hlir/framework/instruction.h (hunk 1 above)
Comment (haozech) [reply cinn/hlir/framework/instruction.h:55]: 因为in_args_和out_args_的第一个元素是原本op的输入输出，当此op被分为多个kernel，即fn_.size() > 1时，in_args_和out_args_会在之前push进每个kernel的输入输出，此时第一个元素已经不需要了，因此删除，让out_args_/in_args_的数量与kernel数量保持一致

File: cinn/hlir/framework/graph_compiler.cc (hunk 2 above)
Comment (haozech) [reply cinn/hlir/framework/graph_compiler.cc]: 已修改，感谢

File: cinn/lang/lower_impl.cc (hunk 3 above)
Comment (haozech) [reply cinn/lang/lower_impl.cc]: 已修改

**Review by wenming2014 — APPROVED** [review]

**Pull request merged**